impersonate/user.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2021 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package impersonate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+func user(ctx context.Context, c Config, client *http.Client) (oauth2.TokenSource, error) {
+	// Default to the longest acceptable value of one hour as the token will
+	// be refreshed automatically if not set.
+	lifetime := 3600 * time.Second
+	if c.Lifetime != 0 {
+		lifetime = c.Lifetime
+	}
+	u := userTokenSource{
+		client:          client,
+		targetPrincipal: c.TargetPrincipal,
+		subject:         c.Subject,
+		lifetime:        lifetime,
+	}
+	u.delegates = make([]string, len(c.Delegates))
+	for i, v := range c.Delegates {
+		u.delegates[i] = formatIAMServiceAccountName(v)
+	}
+	u.scopes = make([]string, len(c.Scopes))
+	copy(u.scopes, c.Scopes)
+	// Don't auto-refresh token if a lifetime is configured.
+	if c.Lifetime != 0 {
+		tok, err := u.Token()
+		if err != nil {
+			return nil, err
+		}
+		return oauth2.StaticTokenSource(tok), nil
+	}
+	return oauth2.ReuseTokenSource(nil, u), nil
+}
+
+type claimSet struct {
+	Iss   string `json:"iss"`
+	Scope string `json:"scope,omitempty"`
+	Sub   string `json:"sub,omitempty"`
+	Aud   string `json:"aud"`
+	Iat   int64  `json:"iat"`
```

Comment:
Should this be `IAt`? Or is it conventionally `Iat`?

Reply:
I always have seen it as `iat` so I would assume `Iat`.